### DISCUSSION TITLE
XML import

### DISCUSSION THREAD

**Discussion by Alexmeb** [totumonline/totum-ru-issues-and-discussions#168] [Как это сделать?] [answered]
Приветствую.
Очень мало информации по импорту из XML, но думаю будет очень востребовано.
Для примера, хотелось бы понять, как реализовать такую задачу, с несколькими полями.

**Файл с примером: [Test.xml.zip](https://github.com/totumonline/totum-ru-issues-and-discussions/files/9159757/Test.xml.zip)**

![Screenshot_1721](https://user-images.githubusercontent.com/109726405/180218326-c89b3c17-0812-454e-8e2e-73513e0bec9e.png)
![Screenshot_1722](https://user-images.githubusercontent.com/109726405/180218918-a0fd5abb-da8a-4051-860c-5f14ed3e8b4b.png)

**Comment by ttm-alexeypolunin** [comment, marked as answer]
В общем XML весьма громоздкая штуковина, со множеством вариантов. Мы конечно выдумали функцию, которая почти все может обработать, но на выходе получается достаточно громоздкий json.

Вот таблица с полем файл, в которое мы кидаем .xml и там по коду действия на изменения он разбирается и кладется в виде json в соседнее поле. Потом уже мы с этим json можем делать что хотим.

<img width="693" alt="Снимок экрана 2022-07-21 в 16 38 01" src="https://user-images.githubusercontent.com/55755565/180227371-fd86ed30-b7f7-4999-921e-b64bb7ad1c13.png">


Код действия с тригерром «При изменении» в поле Файл:

```
= : if(condition: $#nfv != $#lc; then: $set)

set: set(table: $#ntn; field: 'h_razobrano'= $do2; field: $#nf = $#lc)
		do2: xmlExtract(xml: $content; attrpref: "__"; textname: "VALUE")
		content: fileGetContent(file: $#nfv[0][file])
```

+ fileGetContent — получает содержимое файла

+ xmlExtract — разбирает XML


Получается вот такой json:

<img width="943" alt="Снимок экрана 2022-07-21 в 16 39 24" src="https://user-images.githubusercontent.com/55755565/180227782-c4da4ec4-aa42-4cfd-bb60-2f1500712e0d.png">

Что бы где-то кодом достать например эту дату (что на картинке) будет так:

```
= : #h_razobrano[Document][Order][0][Order_Date][0][VALUE]
```

Дальше если вам что-то конкретное из этого куда-то конкретно положить — напишите, мы подскажем.

**Reply by Alexmeb** [reply]
Да, конкретное поле достаётся без проблем.
А можно как-то вытащить список операций (operation) в табличную часть, по типу insert?
Наименование (name) - количество (count) - цена (price)

**Reply by erpic-ru** [reply]
Например, при помощи listreplace в коде действия на изменение с данными: 
```

= :listReplace(list: $#nfv[Document][Operations]; action: $do; value: "value")

do: insert(table: $#ntn;  field: 'name' = $#value[Name][0][VALUE]; field: 'count' = $#value[Count][0][VALUE]; ...)

```

Можно и insertlist приладить при желании.

UDP: Откорректировал под текущий пример (polunin)

**Reply by Alexmeb** [reply]
Мне кажется с этим не будет работать.
![Screenshot_22](https://user-images.githubusercontent.com/109726405/181925704-f61b85d7-8e1a-4901-aaf1-c6f55b4887b7.png)
.

**Reply by ttm-alexeypolunin** [reply]
**Из-за некоторой избыточности xml у нас в этом случае вот такая лесенка:**

<img width="952" alt="Снимок экрана 2022-07-31 в 13 47 07" src="https://user-images.githubusercontent.com/55755565/182022577-3d2c6af6-ebfc-4a34-a796-bb161e773716.png">

Нам надо поднять значение из `VALUE` из вложенного списка на уровень ключа и получить это:

<img width="645" alt="Снимок экрана 2022-07-31 в 13 46 02" src="https://user-images.githubusercontent.com/55755565/182022648-14bd69ac-5503-41b3-878e-8f2b426d215b.png">

С этим уже можно будет работать через insertList.

Делается это через `listReplace`:

```
= : linkToDataJson(title: "test"; data: $data)

data: listReplace(list: #h_razobrano[Document][Order]; action: "Order_Date" = $#row[Order_Date][0][VALUE]; action: "Order_Name" = $#row[Order_Name][0][VALUE]; action: "Order_Count" = $#row[Order_Count][0][VALUE]; action: "Creator_Name" = $#row[Creator_Name][0][VALUE]; action: "Product_Name" = $#row[Product_Name][0][VALUE]; action: "Customer_Name" = $#row[Customer_Name][0][VALUE]; action: "Product_Article" = $#row[Product_Article][0][VALUE]; value: "row")
```

**Reply by Alexmeb** [reply]
Большое спасибо, всё получилось
![Screenshot_23](https://user-images.githubusercontent.com/109726405/182046835-c83d96c2-e9b8-4d28-8d19-67ed52e8003f.png)

**Reply by ttm-alexeypolunin** [reply]
👯‍♀️👯‍♀️👯‍♀️

**Reply by Alexmeb** [reply]
Там правда было пропущено  Operation, помимо Operations, разобрался)
